proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,112 @@
+- This [internal discussion](https://teams.microsoft.com/l/message/19:ed7a508bf00c4b088a7760359f0d0308@thread.skype/1698341652961?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4ba7372f-2799-4677-89f0-7a1aaea3706c&parentMessageId=1698341652961&teamName=.NET%20Developer%20Experience&channelName=InfraSwat&createdTime=1698341652961) from a C# team member. They wanted to use VS as the product is shipped to customers and got blocked when we shipped an SDK that didn't have a corresponding MSI and hence VS couldn't load Roslyn anymore.
+- [VS code](https://github.com/dotnet/vscode-csharp/issues/6471) having to adjust to consider local directories for SDK because our resolver can't find them.
+
+## Detailed Design
+The global.json file will support two new properties under the `sdk` object:
+
+- `"additionalPaths"`: this is a list of paths that the host resolver should consider when looking for compatible SDKs. Relative paths will be interpreted relative to the global.json file.
+- `"additionalPathsOnly"`: when `true` the resolver will _only_ consider paths in `additionalPaths`. It will not consider any machine wide locations (unless they are specified in `additionalPaths`). The default for this property is `false`.
+
+The `additionalPaths` works similar to how multi-level lookup works. It adds additional locations that the host resolver should consider when trying to resolve a compatible .NET SDK. For example:
+
+```json
+{
+    "sdk": {
+        "additionalPaths": [ ".dotnet" ],
+    }
+}
+```
+
+In this configuration the host resolver would find a compatible SDK, if it exists in `.dotnet` or a machine wide location. The host resolver will consider the `additionalPaths` in the order they are defined and will stop at the first match. If none of the locations have a matching SDK then
+ will fall back to the existing SDK resolution strategy: `%PATH%` followed by machine wide installations.
+
+This lookup will stop on the first match which means it won't necessarily find the best match. Consider a scenario with a global.json that has:
+
+```json
+{
+  "sdk": {
+    "additionalPaths": [ ".dotnet" ],
+    "version": "7.0.200",
+    "rollForward": "latestFeature"
+  }
+}
+```
+
+In a scenario where the `.dotnet` directory had 7.0.200 SDK but there was a machine wide install of 7.0.300 SDK, the host resolver would pick 7.0.200 out of `.dotnet`. That location is considered first, it has a matching SDK and hence discovery stops there.
+
+The values in the `additionalPaths` property can be a relative or absolute path. When a relative path is used it will be resolved relative to the location of global.json. These values also support the following substitutions:
+
+- `"$user$"`: this matches the user-local installation point of .NET SDK for the current operating system: `%LocalAppData%\Microsoft\dotnet` on Windows and `$HOME/.dotnet` on Linux/macOS.
```

Comment:
Honestly don't know, could only guess. I would start with the fact that I didn't know where the "user location" was until this doc came about. And also, that `dotnet-install` script wasn't that popular back in the day all of this was designed. Nowadays we seem to have the opposite problem (too many people use the script for things it was not meant to do).

Reply:
> Why aren't user locations included in the default behavior of the resolver already?

I think it would be a security hole. In general, it is not ok for programs installed in admin managed locations to load components from user-controlled location.

Reply:
Good point. Does that potentially points to a hole in this design? If I added `$machine`, `$user` and `.dotnet` to my additional paths, could the machine wide host end up loading a user-installed SDK?

Reply:
If you opt-in into unsecure configuration, it is your problem. You can do that via DOTNET_ROOT today. I think it is important that the defaults are secure.

Reply:
> In general, it is not ok for programs installed in admin managed locations to load components from user-controlled location.

MSBuild does this today via tasks, analyzers, generators, etc ... Is that _okay_ because the user has effectively opted into it by specifying references to the NuPkg that brought down the binaries?